### PR TITLE
Added an early stopping criterion.

### PR DESCRIPTION
**Pull request by maxime1310**
Before stopping the training we check if both the reconstruction loss and the kl divergence have stopped improving over a number of epochs given by the user. 
We look at the relative improvement to determine whether to stop or not.

### CODE REVIEW

File: scvi/train.py [hunk 1]
```diff
@@ -33,15 +44,40 @@ def train(vae, data_loader_train, data_loader_test, n_epochs=20, learning_rate=0
...
+
             real_loss = reconst_loss + kl_divergence
             optimizer.zero_grad()
             train_loss.backward()
             optimizer.step()
 
+        current_performances_kl[:-1] = current_performances_kl[1:]
+        current_performances_reconst[:-1] = current_performances_reconst[1:]
+        current_performances_kl[-1] = total_current_kl.data[0]
+        current_performances_reconst[-1] = total_current_reconst.data[0]
```
Comment (jeff-regier) [inline scvi/train.py]: Why store former values for `reconst_loss` and `kl_divergence` separately, rather than just `train_loss`?

File: scvi/train.py [hunk 2]
```diff
@@ -33,15 +44,40 @@ def train(vae, data_loader_train, data_loader_test, n_epochs=20, learning_rate=0
...
+
+        # Computing the relative improvment of kl and reconstruction loss
+        # over the chosen number of epochs
+        reconst_relative_improvement = ((current_performances_reconst[0] - current_performances_reconst[-1])
+                                        / current_performances_reconst[0])
+        kl_relative_improvement = ((current_performances_kl[0] - current_performances_kl[-1])
+                                   / current_performances_kl[0])
+        # Test whether stopping criterions are met
+        if epoch > patience and kl_relative_improvement < threshold and reconst_relative_improvement < threshold:
```
Comment (jeff-regier) [inline scvi/train.py]: Can we have a single threshold for `total_loss`, rather than separate thresholds for kl and reconst?

File: scvi/train.py (hunk 1 above)
Comment (romain-lopez) [reply scvi/train.py]: Yes, why not look directly at the lower bound ?

File: scvi/train.py (hunk 2 above)
Comment (romain-lopez) [reply scvi/train.py]: agreed, unless there is a specific reason ?

File: scvi/train.py (hunk 1 above)
Comment (maxime1310) [reply scvi/train.py]: small typo I meant reconstruction loss of magnitude  1000 and KLD  of magnitude 10

Comment (jeff-regier) [reply scvi/train.py]: Got it, that makes sense, thanks for the explanation. I'd be good to be more systematic about it eventually, comparing different termination criteria on a lot of datasets. But we can't do that until we have a bunch of datasets available; this is a pretty good solution for now.
